### PR TITLE
Dispatching Python covariance matrices

### PR DESCRIPTION
**Pull request by siranipour**
We now dispatch covmats purely from python. Addresses various points including https://github.com/NNPDF/nnpdf/issues/1476#issuecomment-979823687 and https://github.com/NNPDF/nnpdf/issues/1475#issuecomment-979425362

Let's see what the tests do

### CODE REVIEW

**Comment by scarlehoff** [comment]
I've rebased your commit on top of #1430 to see whether the test passes also there https://github.com/NNPDF/nnpdf/runs/4333530141?check_suite_focus=true

**Comment by siranipour** [comment]
Looks like the linux tests are passing on the rebase

File: validphys2/src/validphys/tests/test_regressions.py [hunk 1]
```diff
@@ -69,15 +69,19 @@ def test_expcovmat(data_config):
 
 @make_table_comp(parse_exp_mat)
 def test_t0covmat(data_witht0_config):
-    return API.groups_covmat_no_table(**data_witht0_config)
+    d = dict(data_witht0_config)
+    d['use_cuts'] = 'internal'
```
Comment (scarlehoff) [inline validphys2/src/validphys/tests/test_regressions.py:73]: You can use this instead: https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/tests/conftest.py#L94

Comment (siranipour) [reply validphys2/src/validphys/tests/test_regressions.py:73]: Wonderful, good spot.

File: validphys2/src/validphys/covmats.py [hunk 2]
```diff
@@ -472,133 +472,10 @@ def groups_corrmat(groups_covmat):
...
-    if norm_threshold is not None:
-        covmat = regularize_covmat(
-            covmat,
-            norm_threshold=norm_threshold
-        )
-    return covmat
-
-
-@check_dataset_cuts_match_theorycovmat
-def covmat(
```
Comment (Zaharid) [inline validphys2/src/validphys/covmats.py:510]: Where has all this code and documentation gone? How are  the `norm_threshold` and `use_weights_in_covmat` arguments handled?

Comment (siranipour) [reply validphys2/src/validphys/covmats.py:510]: The weights are used in the lower level functions https://github.com/NNPDF/nnpdf/blob/eeae19b15c960b4f8f974c3ac2ba2af36ed845bc/validphys2/src/validphys/covmats.py#L35

**Comment by Zaharid** [comment]
I am a bit worried that this is going to obscure `--help` for various things a bit too much. Admittedly that is because reportengine sucks in that regard. That could possibly be remedied with a tutorial on chi²s which we clearly need for other reasons.

**Comment by Zaharid** [comment]
It would be good to have tests that cover the various options passed to the covmat. E.g. the thing with the `norm_threshold` clearly should have failed something.

**Comment by siranipour** [comment]
> I am a bit worried that this is going to obscure `--help` for various things a bit too much. Admittedly that is because reportengine sucks in that regard. That could possibly be remedied with a tutorial on chi²s which we clearly need for other reasons.

Yeah I couldn't figure out how to do this without having a dummy function which the `explicit_node` could dispatch on. Let me know if you can think of a better alternative, I was thinking perhaps I've missed something obvious

> It would be good to have tests that cover the various options passed to the covmat. E.g. the thing with the `norm_threshold` clearly should have failed something.

will add now

**Comment by Zaharid** [comment]
One thing I'd like to do someday is have a more restricted version of `explict_node` that looks like

```
produce_covmat_t0_considered = dispatch_node(on="use_t0", {True: covmats.t0_covmat_from_systematics, False: covmat_from_systematics})
```

which does the same thing but additionally knows  how to follow the possibilities and produce help.

**Comment by siranipour** [comment]
Yes that would be nice, since it could then dispatch another `dispatch_node` which is the problem I was having here.

**Comment by scarlehoff** [comment]
Should I / Can I rebase #1430 on top of this one already or are more changes needed?

**Comment by siranipour** [comment]
I’m done with it if @Zaharid is happy

On 29 Nov 2021, at 10:14, Juan M. Cruz-Martinez ***@***.***> wrote:

﻿

Should I / Can I rebase #1430<https://github.com/NNPDF/nnpdf/pull/1430> on top of this one already or are more changes needed?

**Comment by Zaharid** [comment]
It seems ok. I would hope that the documentation challenges are addressed by #1478.

**Comment by siranipour** [comment]
Shall we merge this?

**Review by scarlehoff — APPROVED** [review]
I'm in favour of merging since everything seems to work in this one and also down the line (#1430 )

**Pull request merged**